core/src/elf_extraction.rs

```diff
@@ -0,0 +1,284 @@
+        Err(_) => {
+            return Err("Error reading ROM file".into());
+        }
+    }
+
+    // Parse the ELF
+    let elf = ElfBytes::<AnyEndian>::minimal_parse(&file_data)?;
+
+    let mut out = ElfPayload { entry_point: elf.ehdr.e_entry, ..Default::default() };
+
+    // Process all section headers
+    if let Some(shdrs) = elf.section_headers() {
+        for sh in shdrs {
+            // Must have bytes in the file (PROGBITS)
+            //
+            // All of the sections we care about are PROGBITS.
+            //
+            // Examples of things this skips are: .bss and any notes
+            if sh.sh_type != SHT_PROGBITS {
+                continue;
+            }
+
+            // Must be allocated at runtime
+            //
+            // Essentially all sections that we need to load into memory when the program is loaded.
+            //
+            // Note: .bss is included in this, but since it is not SHT_PROGBITS, it is never processed.
+            //
+            // Exampple of things this skips are the .debug_* related sections.
+            if (sh.sh_flags & SHF_ALLOC as u64) == 0 {
+                continue;
+            }
+
+            // Spec says ignore if address is 0
+            if sh.sh_addr == 0 {
+                continue;
+            }
+
+            // Grab the section data and word-align it
+            let (raw, _) = elf.section_data(&sh)?;
+            let mut data = raw.to_vec();
+            while data.len() % 4 != 0 {
+                data.pop();
+            }
+
+            // Categorize the section based on its flags
+            let is_exec = (sh.sh_flags & SHF_EXECINSTR as u64) != 0;
+            let is_write = (sh.sh_flags & SHF_WRITE as u64) != 0;
+            let in_ram =
+                sh.sh_addr >= RAM_START_ADDR && sh.sh_addr + data.len() as u64 <= RAM_END_ADDR;
+
+            if is_exec {
+                // Executable code section
+                out.exec.push(DataSection { addr: sh.sh_addr, data });
+            } else if is_write && in_ram {
+                // Read-write data that needs to be copied to RAM
+                out.rw.push(DataSection { addr: sh.sh_addr, data });
+            } else {
+                // Read-only data (constants, strings, etc.)
+                out.ro.push(DataSection { addr: sh.sh_addr, data });
+            }
+        }
+    }
+
+    Ok(out)
+}
+
+/// Helper function to merge adjacent read-only sections
+///
+///   Example: If you have:
+///  - Section A: addr=0x1000, data=[1,2,3,4] (ends at 0x1004)
+///  - Section B: addr=0x1004, data=[5,6,7,8] (starts at 0x1004)
+///
+///  They merge into:
+///  - Single section: addr=0x1000, data=[1,2,3,4,5,6,7,8]
+pub fn merge_adjacent_ro_sections(sections: &[DataSection]) -> Vec<DataSection> {
+    if sections.is_empty() {
+        return Vec::new();
+    }
+
+    let mut merged = Vec::new();
+    let mut sections = sections.to_vec();
+
+    // Sort by address
+    sections.sort_by_key(|s| s.addr);
```

Comment:
Doing the sorting here, once we have all of the sections means:

- This is O(n log n)
- If the sections are out of order, we still don't miss any